### DISCUSSION TITLE
Building C extensions with fpm

### DISCUSSION THREAD

**Discussion by awvwgk** [fortran-lang/fpm#333] [General]
I have recently ported a project to fpm which exports a C-API via `iso_c_binding` (see [`s-dftd3`](https://github.com/awvwgk/simple-dftd3)). I usually put the C headers in `include/` to have them in a prominent place in the repo and export this directory with meson as include-directory.

The project currently fails to build with Fortran fpm because the header cannot be found when compiling the API test:

```
 + gfortran -c test/api/api-test.c -Wall -Wextra -Wimplicit-interface -fPIC -fmax-errors=1 -g -fbounds-check -fcheck-array-temporaries -fbacktrace -fcoarray=single  -J build/gfortran_debug/s-dftd3 -I build/gfortran_debug/s-dftd3 -o build/gfortran_debug/test/api_api-test.c.o
cc1: warning: command-line option ‘-Wimplicit-interface’ is valid for Fortran but not for C
cc1: warning: command-line option ‘-fcheck-array-temporaries’ is valid for Fortran but not for C
cc1: warning: command-line option ‘-fbacktrace’ is valid for Fortran but not for C
cc1: warning: command-line option ‘-fcoarray=single’ is valid for Fortran but not for C
test/api/api-test.c:21:10: fatal error: dftd3.h: No such file or directory
   21 | #include "dftd3.h"
      |          ^~~~~~~~~
compilation terminated.
 Command failed
ERROR STOP
```

Currently my only choice seems to be moving the C API test to somewhere where fpm won't try to compile it and just test the Fortran API.

Is there a way to keep the C API test or do I have to disable this feature for fpm? What would be the correct place to put the header to allow fpm to build the test?

**Comment by LKedward** [comment]
If there aren't objections I'm planning to add support for an `include/` directory in the package root directory which is added to compiler include list for each package.

An inelegant workaround for now could be to make a copy/link of the required header files in the same directory as the test.

**Reply by awvwgk** [reply]
Another issue is that the main is actually a C main not a Fortran one, not sure if this is problematic or not. I decided to disable the C API test for now, but I will revisit this decision when we support include directories.

Sym-linking won't work on Windows, unfortunately and I rather disable a feature than to hack around it.

**Reply by LKedward** [reply]
> Another issue is that the main is actually a C main not a Fortran one, not sure if this is problematic or not

Yeah this isn't supported currently; from your package I see this why this would be useful to add support for.
